core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1743,6 +1762,10 @@ private[spark] class DAGScheduler(
           } else if (mapIndex != -1) {
             // Mark the map whose fetch failed as broken in the map stage
             mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress)
+            if (mapStage.shuffleDep.shuffleMergeEnabled) {
+              mapOutputTracker.
+                unregisterMergeResult(shuffleId, reduceId, bmAddress, Option(mapIndex))
```

Comment:
@otterc one question which I couldn't understand from the internal fix is, should we pass `mapId` or `mapIndex` to the `unregisterMergeResult` here?

Reply:
We need to pass `mapId`. The intention here is to unregister the mergeStatus if the block corresponding to the <shuffleId, mapId, reduceId> got merged for the reduce partition <shuffleId, reduceId>.  mergeStatus tracks `mapIds` and not mapIndices. 
Also we can add a comment here that why are we doing this:
When there is fetch failure for a block <shuffleId, mapId, reduceId> and if that block was merged to partition <shuffleId, reduceId>, it indicates that the iterator also failed to fetch the merged block for <shuffleId, reduceId>. So, we unregister the mergeResult for partition<shuffleId, reduceId> as there is no guarantee that the merged block for <shuffleId, reduceId> from the same blockManager will be successful the next time this stage is resubmitted.

On a side note, this was one of the reasons I thought this change should go with the fetch side change because it makes the explanation easier.

Reply:
@venkata91 I corrected the above response. Passing mapIndex is correct, because that is what we send in the ShuffleBlockPush message.

Reply:
Thanks. Added a comment to keep it clear as well.